Add permission prompt related docs (#85)
WICG/private-network-access: permission_prompt/security_privacy_self_review.md

## 2. Do features in your specification expose the minimum amount of information necessary to enable their intended uses? 

Yes, apart from the above.

## 3. How do the features in your specification deal with personal information, personally-identifiable information (PII), or information derived from them?

It does not.

## 4. How do the features in your specification deal with sensitive information?

It does not.

## 5. Do the features in your specification introduce new state for an origin that persists across browsing sessions?

Yes, the permission status which allows a certain public webpage to access a certain private/local service.

## 6. Do the features in your specification expose information about the underlying platform to origins?

See question 1.

## 7. Does this specification allow an origin to send data to the underlying platform?

No.

## 8. Do features in this specification enable access to device sensors?

No.

## 9. Do features in this specification enable new script execution/loading mechanisms?

No.

## 10. Do features in this specification allow an origin to access other devices?

> No, instead it restricts the ability of origins to access local network devices.

Reviewer: Fair point, we should be more nuanced here. This explainer opens up access to devices from new contexts, as compared against a baseline of the previous version of Private Network Access. However, as compared to the current state of the web, it mostly introduces new hurdles (a permission prompt) instead of granting new capabilities. I say mostly, because it does grant new capabilities to secure contexts. These, however, are removed at the same time from non-secure contexts, which is an overall win: the capability has become restricted to secure contexts.

> No, instead it restricts the ability of origins to access local network devices.

Author: Changed the statements.

## 11. Do features in this specification allow an origin some measure of control over a user agent’s native UI?

No.

## 12. What temporary identifiers do the features in this specification create or expose to the web?

None.

> ## 13. How does this specification distinguish between behavior in first-party and third-party contexts?

Reviewer 2: This should probably also discuss how the permission prompt can only be presented by first parties?

Reviewer: Good idea! We should restrict usage of this API and permission to first parties (i.e. add this to the explainer) and mention it here.

Author: Good point here. I would suggest that @annevk the "first parties" you are talking here is about the permission policies. My idea is that we can allow both top-level document and same-origin frames be able to raise up the permission by default. In the future, we might be able to allow top-level document delegating the permission to third-party frames. WDYT?

Reviewer: I think we landed on top-level documents only at first, suggested an edit below to clarify that.

Author: Yes, what @letitz said is correct. We did an offline discuss and decide to land on top-level documents only first. However, we are still open for options here.

Third-party iframes are treated distinctly from the first party embedder: even
if the first party is served from non-public address space, only the third
party's address space is considered when applying private network request checks

> to requests made by the third-party iframe.

Reviewer: We should not allow access to this API in third-party iframes.

Third-party scripts are not treated distinctly. In the case of
`document.addressSpace`, this could reveal some information about network
configuration to third parties. There seems to be no particular reason to treat
third party scripts differently when checking outgoing requests, however.

One area of discussion is
whether or not to treat sandboxed iframes as `public`: see [#26](https://github.com/WICG/private-network-access/issues/26). It seems a good
idea to do so, enabling web developers to include third-party content on
non-public websites without allowing it to poke at non-public resources. On the
other hand, this might help malicious websites trying to determine which IP
address they are being accessed from by providing them with a baseline against

> which to compare their own capabilities.

Reviewer: I think this can be removed if we restrict to first-party iframes?

## 10. Do features in this specification allow an origin to access other devices?

> Compare to the previous version of Private Network Access, this feature relaxes

Reviewer: ```suggestion
Compared to the previous version of Private Network Access, this feature relaxes
```

## 13. How does this specification distinguish between behavior in first-party and third-party contexts?

> For permission prompt in particular, it would be restricted to first-party
> embedders in current stage.

Reviewer: ```suggestion
The permission can only be granted to the top-level page.
```

> We might want to relax the boundary and also allow same-origin child context to
> be able to raise the permission by default. However, third-party iframes would
> be exclude from that.

Reviewer: ```suggestion
We might want to relax the boundary and also allow same-origin child contexts to
be able to request the permission by default. However, third-party contexts (e.g. iframes) would
not be allowed.
```

> In the future, it might make sense to allow top-level documents to control its
> permission with permission policy and even allow the private network access
> permission to be delegated to third-party contexts. In that case, third-party
> context should be the one taking into account for the permission.

Reviewer: ```suggestion
In the future, it might make sense to allow top-level documents to control its
permission with permission policy and even allow the private network access
permission to be delegated to third-party contexts. In that case, the third-party
context should be the one taken into account for the permission.
```
